src/rvsdg/restructure.rs

```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
                     .graph
-                    .neighbors_directed(ix, Direction::Incoming)
-                    .filter(|pred| !dominates(ix, *pred))
-                    .nth(1)
-                    .is_some()
+                    .edges_directed(*node, Direction::Outgoing)
+                    .filter(|x| !dgraph.contains(&x.target()))
                 {
-                    tail_continuations.entry(idom).or_default().push(ix);
+                    cont.exit_arcs
+                        .entry(edge_ref.id())
+                        .or_default()
+                        .insert(exit_edge.id());
+                    cont.reentry_nodes.insert(exit_edge.target());
                 }
             }
         }
+        cont
+    }
+    fn traverse_linear_region(&self, mut cur: NodeIndex, exit: NodeIndex) -> NodeIndex {
+        while cur != exit {
+            let mut neighbors = self.graph.neighbors_directed(cur, Direction::Outgoing);
+            let Some(next) = neighbors.next() else {
+                break;
+            };
+            if neighbors.next().is_some() {
+                break;
+            }
+            cur = next;
+        }
+        cur
+    }
+    fn divert_edge(&mut self, edge: EdgeIndex, new_target: NodeIndex) -> EdgeIndex {
+        let (src, target) = self.graph.edge_endpoints(edge).unwrap();
+        if target == new_target {
+            return edge;
+        }
+        let weight = self.graph.remove_edge(edge).unwrap();
+        self.graph.add_edge(src, new_target, weight)
+    }
+    fn restructure_branches_inner(
+        &mut self,
+        entry: NodeIndex,
+        exit: NodeIndex,
+        state: &mut RestructureState,
+    ) {
+        // Use the original algorithm (RVSDG Paper, including the accompanying JLM code).
+        let start = self.traverse_linear_region(entry, exit);
+        if start == exit {
+            return;
+        }
+        let cont = self.get_continuation(start);
 
-        for (idom, conts) in tail_continuations.into_iter() {
-            // Split any direct edges from the idom to the continuation. If we
-            // have more than one continuation then split _all_ edges.
-            for cont in &conts {
-                let mut walker = self
-                    .graph
-                    .neighbors_directed(*cont, Direction::Incoming)
-                    .detach();
-                while let Some((e, src)) = walker.next(&self.graph) {
-                    if src == idom && conts.len() == 1 {
-                        self.split_arc(e);
-                    } else if conts.len() > 1 {
-                        self.split_edges(src);
-                    }
+        if cont.reentry_nodes.is_empty() {
+            // Nothing to do.
+            return;
+        }
+        // assert!(
+        //     !cont.reentry_nodes.is_empty(),
+        //     "No reentry nodes found [start: {start:?}, entry: {entry:?}, exit: {exit:?}]"
+        // );
+
+        if cont.reentry_nodes.len() == 1 {
```

Comment:
Which of these if statements are "optimizations" that generate better cfgs and which ones are necessary? For example, is this one necessary or does the else case technically work?

Reply:
Added some comments here. This structure is also in the JLM repo: it avoids us adding extra nodes and extra predicates when we don't need to.